### ISSUE TITLE
Question about saving meta model

### ISSUE THREAD

**Issue by FLT19940317** [zhuhancheng/MetaIQA#5]
[https://github.com/zhuhancheng/MetaIQA/blob/master/MetaIQA_On_TID2013_KADID.py#L459](url)
`best_model = copy.deepcopy(model)`
To my understanding, Should here be that test meta_model and save meta_model?

**Comment by zhuhancheng** [comment]
In the code, **meta_model** is only a temporary model. Our optimization and testing are based on the **model**.

**Comment by ritbrz** [comment]
I also have same confusion that it should save `meta_model` instead of `model`.

In the last five `list_noise` , `model` will execute 5*2 times `optimizer.step()` , and the `meta_model` is updated by the diff of `model` and `temp_model`.

When to save best model, `best_model = copy.deepcopy(model)` will save the `model` after  5*2 times  `optimizer.step()` instead of last `meta_model` computed by `name_to_param1[name].data.add_(diff / task_num)`. 

So, finally, the best model saved actually have finished fine-tuning by last five `list_noise` , rather than the meta model contains meta-knowledge.

:( I'm not sure I understand this correctly.

**Comment by FLT19940317** [comment]
> I also have same confusion that it should save `meta_model` instead of `model`.
> 
> In the last five `list_noise` , `model` will execute 5*2 times `optimizer.step()` , and the `meta_model` is updated by the diff of `model` and `temp_model`.
> 
> When to save best model, `best_model = copy.deepcopy(model)` will save the `model` after 5*2 times `optimizer.step()` instead of last `meta_model` computed by `name_to_param1[name].data.add_(diff / task_num)`.
> 
> So, finally, the best model saved actually have finished fine-tuning by last five `list_noise` , rather than the meta model contains meta-knowledge.
> 
> :( I'm not sure I understand this correctly.

**Comment by FLT19940317** [comment]
> I also have same confusion that it should save `meta_model` instead of `model`.
> 
> In the last five `list_noise` , `model` will execute 5*2 times `optimizer.step()` , and the `meta_model` is updated by the diff of `model` and `temp_model`.
> 
> When to save best model, `best_model = copy.deepcopy(model)` will save the `model` after 5*2 times `optimizer.step()` instead of last `meta_model` computed by `name_to_param1[name].data.add_(diff / task_num)`.
> 
> So, finally, the best model saved actually have finished fine-tuning by last five `list_noise` , rather than the meta model contains meta-knowledge.
> 
> :( I'm not sure I understand this correctly.

I have the same thought with you.

**Comment by zhuzhu18** [comment]
> I also have same confusion that it should save `meta_model` instead of `model`.
> 
> In the last five `list_noise` , `model` will execute 5*2 times `optimizer.step()` , and the `meta_model` is updated by the diff of `model` and `temp_model`.
> 
> When to save best model, `best_model = copy.deepcopy(model)` will save the `model` after 5*2 times `optimizer.step()` instead of last `meta_model` computed by `name_to_param1[name].data.add_(diff / task_num)`.
> 
> So, finally, the best model saved actually have finished fine-tuning by last five `list_noise` , rather than the meta model contains meta-knowledge.
> 
> :( I'm not sure I understand this correctly.

I agree with you. And I have another problem, what's the use of temp_model?
